### PR TITLE
[GLUTEN-4668][CH] Merge two phase hash-based aggregate into one aggregate in the spark plan when there is no shuffle

### LINKED ISSUE #4668: [CH] Merge two phase hash-based aggregate into one aggregate in the spark plan when there is no shuffle
### Description

Merge two phase hash-based aggregate into one aggregate in the spark plan when there is no shuffle:
Examples:
```
 HashAggregate(t1.i, SUM, final)
                |                                 =>    HashAggregate(t1.i, SUM, complete)
 HashAggregate(t1.i, SUM, partial)
```
now this feature only support for CH backend.

### PR DESCRIPTION
**Pull request by zzcclp**
## What changes were proposed in this pull request?

Merge two phase hash-based aggregate into one aggregate in the spark plan when there is no shuffle between them:
Examples:
```
 HashAggregate(t1.i, SUM, final)
                |                  =>    HashAggregate(t1.i, SUM, complete)
 HashAggregate(t1.i, SUM, partial)
```
For example:
TPCH Q18 with bucket tables, before this pr:
tmp2
there are two `HashAggregateTransformer` in one whole stage;

after this pr:
tmp1
there is only one `HashAggregateTransformer` in one whole stage, and will reduce the time for the second `HashAggregateTransformer`.

**Now this feature only support for CH backend.**

Close #4668.

(Fixes: #4668)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

**Comment by zzcclp** [comment]
@liujiayi771 @lgbo-ustc please help to review, thanks.

File: gluten-core/src/main/scala/io/glutenproject/extension/MergeTwoPhasesHashAggregate.scala
Comment (liujiayi771) [inline gluten-core/src/main/scala/io/glutenproject/extension/MergeTwoPhasesHashAggregate.scala:51]: It is better to use `PhysicalPlanSelector.maybe` to check if guten is enabled.

File: gluten-core/src/main/scala/io/glutenproject/extension/MergeTwoPhasesHashAggregate.scala
Comment (liujiayi771) [inline gluten-core/src/main/scala/io/glutenproject/extension/MergeTwoPhasesHashAggregate.scala:73]: Can we use
```scala
hashAgg.copy(
  aggregateExpressions = completeAggregateExpressions,
  child = child.child
)
```

File: gluten-core/src/main/scala/io/glutenproject/extension/MergeTwoPhasesHashAggregate.scala
Comment (liujiayi771) [reply gluten-core/src/main/scala/io/glutenproject/extension/MergeTwoPhasesHashAggregate.scala:51]: Perhaps you should check whether the `SparkPlan` has been tagged as `TRANSFORM_UNSUPPORTED`. There are some rules earlier that will tag the `SparkPlan`, for example, `FallbackOnANSIMode` will tag all `SparkPlan`s as `TRANSFORM_UNSUPPORTED` when ANSI mode is enabled. In this case, we can also avoid merging the aggregations.

Comment (zzcclp) [reply gluten-core/src/main/scala/io/glutenproject/extension/MergeTwoPhasesHashAggregate.scala:51]: done

File: gluten-core/src/main/scala/io/glutenproject/extension/MergeTwoPhasesHashAggregate.scala
Comment (zzcclp) [reply gluten-core/src/main/scala/io/glutenproject/extension/MergeTwoPhasesHashAggregate.scala:73]: done

**Comment by zzcclp** [comment]
@liujiayi771 @lgbo-ustc @ulysses-you @PHILO-HE @rui-mo    please help to review again, thanks.

File: gluten-core/src/main/scala/io/glutenproject/extension/MergeTwoPhasesHashAggregate.scala
Comment (ulysses-you) [inline gluten-core/src/main/scala/io/glutenproject/extension/MergeTwoPhasesHashAggregate.scala:99]: why not put this `if` into previous line ?

File: gluten-core/src/main/scala/io/glutenproject/extension/MergeTwoPhasesHashAggregate.scala
Comment (ulysses-you) [inline gluten-core/src/main/scala/io/glutenproject/extension/MergeTwoPhasesHashAggregate.scala:136]: can we also handle SortAggregate ? it is possible that there is no shuffle and sort between two SortAggregate.

File: backends-clickhouse/src/main/scala/io/glutenproject/execution/CHHashAggregateExecTransformer.scala
Comment (ulysses-you) [inline backends-clickhouse/src/main/scala/io/glutenproject/execution/CHHashAggregateExecTransformer.scala:121]: I'm not sure how CH backend transform aggregate. But the code seems different with before, `Partial` can appear with `PartialMerge` with distinct.

**Review by rui-mo — COMMENTED** [review]
Could you add some description for the motivation of this PR? It is a usual case for partial + final without shuffle between? Thanks.

File: gluten-core/src/main/scala/io/glutenproject/extension/MergeTwoPhasesHashAggregate.scala
Comment (zzcclp) [reply gluten-core/src/main/scala/io/glutenproject/extension/MergeTwoPhasesHashAggregate.scala]: done

File: gluten-core/src/main/scala/io/glutenproject/extension/MergeTwoPhasesHashAggregate.scala
Comment (zzcclp) [reply gluten-core/src/main/scala/io/glutenproject/extension/MergeTwoPhasesHashAggregate.scala:154]: added SortAggregate

File: backends-clickhouse/src/main/scala/io/glutenproject/execution/CHHashAggregateExecTransformer.scala
Comment (zzcclp) [reply backends-clickhouse/src/main/scala/io/glutenproject/execution/CHHashAggregateExecTransformer.scala]: reverted

**Comment by zzcclp** [comment]
> Could you add some description for the motivation of this PR? It is a usual case for partial + final without shuffle between? Thanks.

Updated, please review again, thanks.  @ulysses-you @rui-mo

File: gluten-core/src/main/scala/io/glutenproject/extension/MergeTwoPhasesHashAggregate.scala
Comment (ulysses-you) [inline gluten-core/src/main/scala/io/glutenproject/extension/MergeTwoPhasesHashAggregate.scala:60]: Here we have not pulled out pre/post project, so this validation seems very likely to fail. Do we need this check ? I think it should be fine even we fallback to vanilla Spark after merging aggregates, since there already exists a rule ReplaceHashWithSortAgg to do the similar thing.

Comment (zzcclp) [reply gluten-core/src/main/scala/io/glutenproject/extension/MergeTwoPhasesHashAggregate.scala:60]: I am not sure that merging sort aggregates also works for the vanilla Spark, do you have any idea for this? the hash based aggregates may be work even fallback.

Comment (zzcclp) [reply gluten-core/src/main/scala/io/glutenproject/extension/MergeTwoPhasesHashAggregate.scala]: removed

**Comment by ulysses-you** [comment]
lgtm if test pass, cc @rui-mo if you have other comments

**Review by rui-mo — APPROVED** [review]

**Comment by lgbo-ustc** [comment]
LGTM

**Review by ulysses-you — APPROVED** [review]

**Pull request merged**
